plugins/influxdb/influxdb.go

```diff
@@ -0,0 +1,142 @@
+package influxdb
+
+import (
+	"bytes"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/DataDog/datadog-go/statsd"
+	"github.com/Sirupsen/logrus"
+	"github.com/stripe/veneur/plugins"
+	"github.com/stripe/veneur/samplers"
+)
+
+var _ plugins.Plugin = &InfluxDBPlugin{}
+
+// InfluxDBPlugin is a plugin for emitting metrics to InfluxDB.
+type InfluxDBPlugin struct {
+	Logger     *logrus.Logger
+	InfluxURL  string
+	HTTPClient *http.Client
+	Statsd     *statsd.Client
+}
+
+// NewInfluxDBPlugin creates a new Influx Plugin.
+func NewInfluxDBPlugin(logger *logrus.Logger, addr string, consistency string, db string, client *http.Client, stats *statsd.Client) *InfluxDBPlugin {
+	plugin := &InfluxDBPlugin{
+		Logger:     logger,
+		HTTPClient: client,
+		Statsd:     stats,
+	}
+
+	inurl, err := url.Parse(addr)
+	if err != nil {
+		logger.Fatalf("Error parsing URL for InfluxDB: %q", err)
+	}
+
+	// Construct a path we will be using later.
+	inurl.Path = "/write"
+	q := inurl.Query()
+	q.Set("db", db)
+	q.Set("precision", "s")
+	inurl.RawQuery = q.Encode()
+	plugin.InfluxURL = inurl.String()
+
+	return plugin
+}
+
+// Flush sends a slice of metrics to InfluxDB
+func (p *InfluxDBPlugin) Flush(metrics []samplers.DDMetric, hostname string) error {
+	p.Statsd.Gauge("flush.post_metrics_total", float64(len(metrics)), nil, 1.0)
+	// Check to see if we have anything to do
+	if len(metrics) == 0 {
+		p.Logger.Info("Nothing to flush, skipping.")
+		return nil
+	}
+
+	buff := bytes.Buffer{}
+	colons := regexp.MustCompile(":")
+	for _, metric := range metrics {
+		tags := strings.Join(metric.Tags, ",")
+		cleanTags := colons.ReplaceAllLiteralString(tags, "=")
+		buff.Write(
+			[]byte(fmt.Sprintf("%s,%s value=%f %d\n", metric.Name, cleanTags, metric.Value[0][1], int64(metric.Value[0][0]))),
+		)
+	}
+
+	p.postHelper(p.InfluxURL, &buff)
+
+	return nil
+}
+
+// Name returns the name of the plugin.
+func (p *InfluxDBPlugin) Name() string {
+	return "influxdb"
+}
+
+// Common for POSTing to an endpoint, that consumes JSON.
+func (p *InfluxDBPlugin) postHelper(endpoint string, bodyBuffer *bytes.Buffer) error {
```

Comment:
Let's make the body an `io.Reader` - that'll make it easier for us to test this helper independently.

At first, it might look like you can't do that, because you use the `Len()` method (to report the payload size). There are two ways of getting around that. The easier way, which will work for production but not always report the size in tests (which is fine), is:

```go
 func (p *InfluxDBPlugin) postHelper(endpoint string, body io.Reader) error { 
    type lengther interface{
        Len() int
    }
    if lengther, ok := bodyBuffer.(lengther); ok {
         p.Statsd.Histogram("influxdb_post.content_length_bytes", float64(lengther.Len()), nil, 1.0) 
    }
   ...
}
```

The other approach would be to use a local reader that tees the output and counts up the size. I wouldn't recommend that here, because it's overkill, but it's what you might want to do if we were calling `postHelper` from multiple places and absolutely always needed the length each time.